scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,601 @@
+"""
+gensa: A generalized simulated annealing global optimization algorithm
+"""
+from __future__ import division, print_function, absolute_import
+
+import numpy as np
+import time
+from scipy.optimize import OptimizeResult
+from scipy.optimize import _lbfgsb
+from scipy.special import gammaln
+
+__all__ = ['gensa']
+
+KSPRING = 1.e8
+BIG_VALUE = 1.e13
+MAX_REINIT_COUNT = 100
+
+
+class GenSARunnerException(Exception):
+    pass
+
+
+class GenSARunner(object):
+    """This class implements the core of the gensa algorithm.
+
+    x0 : ndarray
+        The starting coordinates.
+    fun : callable
+        The objective function
+    lower: ndarray
+        The lower bounds of the search domain
+    upper: ndarray
+        The upper bounds of the search domain
+    fargs: dict
+        Additional arguments to pass to the objective function
+
+    """
+
+    def __init__(self, fun, x0, lower, upper, fargs=()):
+        self._lower = np.array(lower)
+        self._upper = np.array(upper)
+        assert(len(lower) == len(
+            upper)),'Lower and upper bounds are inconsistent'
+        if x0 is None:
+            x0 = self._lower + np.random.rand(len(self._lower)) * (
+                    self._upper - self._lower)
+        else:
+            x0 = np.array(x0)
+        assert(len(self._lower) == len(
+            x0)), 'Lower bounds size does not match x0'
+        assert(len(self._upper) == len(
+            x0)), 'Upper bounds size does not match x0'
+        self._fun = fun
+        self._fargs = fargs
+        self._x = np.array(x0)
+        self._dim, = self._x.shape
+        self._xrange = self._upper - self._lower
+        self._xbackup = np.array(self._x)
+        self._xmini = np.array(self._x)
+        self._xbuffer = np.zeros(self._dim)
+        self._itsoftmax = self._dim * 6
+        self._factr = 1000
+        self._pgtol = 1.e-6
+        self._reps = 1.e-6
+        self._markovlength = self._dim * 2
+        self._hasconstraint = False
+        self._tempsta = 5230
+        self._temp = self._tempsta
+        self._knowreal = False
+        self._maxtime = 3600
+        self._maxfuncall = 1e7
+        self._maxsteps = 500
+        self._temprestart = 0.1
+        self._qv = 2.62
+        self._usey = 1
+        self._xgas = 0.0
+        self._ygas = 0.0
+        self._ranbyx = 0.0
+        self._ranbyy = 0.0
+        self._sgas = -1.0
+        self._qa = -5.0  # Acceptance parameter
+        self.nblocal = 0
+        self._step_record = 0
+        self._nbfuncall = 0
+        self._extensive = False
+
+    def _judge_constraint(self):
+        # Not implemented yet
+        return True
+
+    def initialize(self):
+        """
+        Random coordinate generation in case given initial coordinates
+        are giving invalid objective value
+        """
+        self._nbfuncall = 0
+        self._indTrace = 0
+        if self._markovlength % self._dim != 0:
+            raise GenSARunnerException('Incorrect markov length.')
```

Comment:
Probably not necessary to create a custom exception. Can't we just use `ValueError`, etc?